pkg/csi/service/vanilla/controller.go

```diff
@@ -2274,6 +2267,291 @@ func (c *controller) ListSnapshots(ctx context.Context, req *csi.ListSnapshotsRe
+				"failed to check if cns snapshot is supported on VC %s due to error: %v", vCenterHost, err)
+		}
+		if !isCnsSnapshotSupported {
+			return nil, "", logger.LogNewErrorCodef(log, codes.Unimplemented,
+				"VC %s version does not support snapshot operations", vCenterHost)
+		}
+		// Validate that the volume-id is of block volume type.
+		queryFilter := cnstypes.CnsQueryFilter{
+			VolumeIds: []cnstypes.CnsVolumeId{{Id: volumeID}},
+		}
+		querySelection := cnstypes.CnsQuerySelection{
+			Names: []string{string(cnstypes.QuerySelectionNameTypeVolumeType)},
+		}
+		queryResult, err := utils.QueryVolumeUtil(ctx, volManager, queryFilter, &querySelection, true)
+		if err != nil {
+			return nil, "", logger.LogNewErrorCodef(log, codes.Internal,
+				"queryVolumeUtil failed for vCenter %s with err=%+v", vCenterHost, err)
+		}
+
+		if len(queryResult.Volumes) == 0 {
+			return nil, "", logger.LogNewErrorCodef(log, codes.Internal,
+				"volumeID %q not found in QueryVolumeUtil in vCenter %s", volumeID, vCenterHost)
+		}
+		if queryResult.Volumes[0].VolumeType == common.FileVolumeType {
+			return nil, "", logger.LogNewErrorCodef(log, codes.Unimplemented,
+				"ListSnapshot for file volume: %q not supported", volumeID)
+		}
+		// Note: Re-using original Query Snapshot function since we expect snapshots reside in same VC as source vol,
+		//		where pagination gets handled on the query results from single VC by underlying query function.
+		return common.QueryVolumeSnapshotsByVolumeID(ctx, volManager, volumeID, maxEntries)
+	} else {
+		// Fetch all snapshot entries from all vCenters configured and filter them as per request
+		return queryAllVolumeSnapshotsVanilla(ctx, c, token, maxEntries)
+	}
+}
+
+// queryAllVolumeSnapshotsVanilla(): This func fetches all volume snapshots from all vCenters configured
+// and filters the result as per token provided
+func queryAllVolumeSnapshotsVanilla(ctx context.Context, c *controller, token string,
+	maxEntries int64) ([]*csi.Snapshot, string, error) {
+	var (
+		err                 error
+		csiSnapshots        []*csi.Snapshot
+		nextToken           string
+		snapshotQueryFilter cnstypes.CnsSnapshotQueryFilter
+		finalSnapEntries    []cnstypes.CnsSnapshotQueryResultEntry
+		snapQueryEntries    = make([]cnstypes.CnsSnapshotQueryResultEntry, 0)
+	)
+	log := logger.GetLogger(ctx)
+	startingToken := 0
+	if token != "" {
+		startingToken, err = strconv.Atoi(token)
+		if err != nil {
+			log.Errorf("failed to parse the token: %s err: %v", token, err)
+			return nil, "", err
+		}
+	}
+	// Get all snapshot entries from available vCenters
+	// startingToken = 0 indicates listSnapshots called as a new request and not part of
+	// previous request. Hence fetch all snapshots from CNS
+	if startingToken == 0 || len(CNSSnapshotsForListSnapshots) == 0 {
+		// For multi-VC configuration, query volumes from all vCenters
+		vCenterManager := getVCenterManagerForVCenter(ctx, c)
+		if multivCenterCSITopologyEnabled {
```

Comment:
If this is not enabled then wouldn't:
```
CNSSnapshotsForListSnapshots = snapQueryEntries
CNSVolumeDetailsMap = cnsVolumeDetailsMap
```
Return empty results?

Reply:
when multivCenterCSITopologyEnabled is not enabled, it would be single VC case, which should query and return snapshots from that single VC configured.